bugfix: Make `CCheckQueue` RAII-styled (attempt 2)

hebasto opened the pull request:
This PR:
- makes `CCheckQueue` RAII-styled
- gets rid of the global `scriptcheckqueue`
- fixes https://github.com/bitcoin/bitcoin/issues/25448

The previous attempt was in https://github.com/bitcoin/bitcoin/pull/18731.

w0xlt reviewed (commented):
Approach ACK

ajtowns commented on src/checkqueue.h:
```diff
@@ -143,6 +143,9 @@ class CCheckQueue
     }
 
+    CCheckQueue(const CCheckQueue&) = delete;
+    CCheckQueue& operator=(const CCheckQueue&) = delete;
```
Should apply rule-of-5 and default the move operations, no?

hebasto replied:
Thanks! [Done](https://github.com/bitcoin/bitcoin/pull/26762#issuecomment-1369886008).

ajtowns reviewed (commented):
Concept ACK

hebasto commented:
Updated 2d248d6b80147e5f9f62f4c543d4dbb10c0d95c7 -> 5a19c3962c8702ba3b17f986524e1eb5de049c0e ([pr26762.01](https://github.com/hebasto/bitcoin/commits/pr26762.01) -> [pr26762.02](https://github.com/hebasto/bitcoin/commits/pr26762.02), [diff](https://github.com/hebasto/bitcoin/compare/pr26762.01..pr26762.02)):

- addressed @ajtowns's comments

hebasto commented on src/validation.cpp:
```diff
@@ -1492,6 +1492,7 @@ Chainstate::Chainstate(
     ChainstateManager& chainman,
     std::optional<uint256> from_snapshot_blockhash)
     : m_mempool(mempool),
+      m_script_check_queue{std::make_unique<CCheckQueue<CScriptCheck>>(/*batch_size=*/128, chainman.m_options.worker_threads_num)},
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/26762#issuecomment-1369886008).

hebasto commented:
Rebased 5a19c3962c8702ba3b17f986524e1eb5de049c0e -> b429f92499c0d0224f6be6afae88fe57743bc2de ([pr26762.02](https://github.com/hebasto/bitcoin/commits/pr26762.02) -> [pr26762.03](https://github.com/hebasto/bitcoin/commits/pr26762.03)) due to thr conflict with #26251.

ajtowns commented on src/node/chainstatemanager_args.cpp:
```diff
@@ -34,6 +36,15 @@ std::optional<bilingual_str> ApplyArgsManOptions(const ArgsManager& args, Chains
+        // -par=-n means "leave n cores free" (number of cores - n - 1 script threads)
+        par_value += GetNumCores();
+    }
+    opts.worker_threads_num = std::clamp(par_value - 1, 0, MAX_SCRIPTCHECK_THREADS);
```
Missing the comment explaining the `- 1`

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/26762#issuecomment-1408502815).

ajtowns commented on src/validation.cpp:
```diff
@@ -5307,10 +5300,17 @@ static ChainstateManager::Options&& Flatten(ChainstateManager::Options&& opts)
+    : m_script_check_queue{/*nBatchSizeIn=*/128},
+      m_options{Flatten(std::move(options))}
+{
+    m_script_check_queue.StartWorkerThreads(m_options.worker_threads_num);
```
Previously this was conditional on `worker_threads_num > 0` (since `StartScriptCheckWorkerThreads` would not be called unless `script_threads >= 1`). I think this is okay, since in that case the function only sets some variables that will never be used and skips over a loop.

hebasto replied:
This code will gone in the "refactor: Make `CCheckQueue` constructor start worker threads" commit.

ajtowns commented:
utACK f0209248e1baea41aae72147c973f1f5aa27c97b

Seems reasonable to me. Not 100% confident over delaying stopping the threads until the destructor; but also about equally unsure about how it works now...

ajtowns commented:
> add a default initializer for `kernel::ChainstateManagerOpts::worker_threads_num`

Yikes. Work on adding [`cppcoreguidelines-pro-type-member-init`](https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/pro-type-member-init.html)` to clang-tidy maybe?

reACK 5a7932f395c675fad332cbcd0498bb9fefcb33e0

sedited reviewed (commented):
Concept ACK

hebasto commented:
> Yikes. Work on adding [`cppcoreguidelines-pro-type-member-init`](https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/pro-type-member-init.html)` to clang-tidy maybe?

Well, it fires too many false warnings for our code base (for example, for `CTxOut::CTxOut()` constructor).

hebasto commented:
Updated 5a7932f395c675fad332cbcd0498bb9fefcb33e0 -> f1370b2c1586f7fe487d9f17ee53bcd9b87a9f23 ([pr26762.04](https://github.com/hebasto/bitcoin/commits/pr26762.04) -> [pr26762.05](https://github.com/hebasto/bitcoin/commits/pr26762.05)):

- rebased
- addressed some of @TheCharlatan's comments

hebasto commented on src/node/chainstatemanager_args.cpp:
```diff
@@ -34,6 +36,16 @@ std::optional<bilingual_str> ApplyArgsManOptions(const ArgsManager& args, Chains
     if (auto value{args.GetIntArg("-maxtipage")}) opts.max_tip_age = std::chrono::seconds{*value};
 
+    int par_value = args.GetIntArg("-par", DEFAULT_SCRIPTCHECK_THREADS);
+    if (par_value <= 0) {
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/26762#issuecomment-1479655825).

hebasto commented on src/node/chainstatemanager_args.cpp:
```diff
@@ -34,6 +36,16 @@ std::optional<bilingual_str> ApplyArgsManOptions(const ArgsManager& args, Chains
 
     if (auto value{args.GetIntArg("-maxtipage")}) opts.max_tip_age = std::chrono::seconds{*value};
 
+    int par_value = args.GetIntArg("-par", DEFAULT_SCRIPTCHECK_THREADS);
```
> This is not consistent with how the other options are handled. This overrides the value of the passed in option even if no argument was passed in. I would handle this the same way as `max_tip_age`...

I'm not sure about this change considering this PR scope. The semantic of the "-par" and "-maxtipage" options are quite different.

> ... and also move the `MAX_SCRIPTCHECK_THREADS` to `chainstatemanager_opts.h`.

Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/26762#issuecomment-1479655825).

martinus commented on src/checkqueue.h:
```diff
@@ -181,24 +174,16 @@ class CCheckQueue
+    ~CCheckQueue()
     {
-        WITH_LOCK(m_mutex, m_request_stop = true);
+        m_request_stop = true;
```
I think that removing the lock here could cause a deadlock.
`m_request_stop` is set to true and then `notify_all()` is called; but when there is a worker thread currently standing inside the `while (queue.empty() && !m_request_stop) {` loop but before the `cond.wait(lock);` it will not receive the notification and then proceed to wait in the `cond` indefinitely.

That deadlock was not possible before when the `m_mutex` had to be acquired to change `m_request_stop`, because then it was guaranteed that all threads were either waiting on the condition variable or waiting outside that loop to get the lock

hebasto replied:
Doesn't the default `memory_order_seq_cst` memory order for `std::atomic<bool> m_request_stop` guarantee the correct behaviour?

martinus replied:
The only thing guaranteed is the order; so `m_request_stop` will be set to `true` before `notify_all` is called. But there is no guarantee where the other threads currently are because there is now no lock anymore. E.g. a worker thread could be currently here:

```cpp
while (queue.empty() && !m_request_stop) {
    if (fMaster && nTodo == 0) {   // <-------- Worker thread waiting here due to context switch
        nTotal--;
        bool fRet = fAllOk;
        // reset the status for new work later
        fAllOk = true;
        // return the current status
        return fRet;
    }
    nIdle++;
    cond.wait(lock); // wait
    nIdle--;
}
```

Now while the worker thread is not running for whatever reason (context switch, high CPU load, ...) the main thread runs, calls `m_request_stop = true`, then `m_worker_cv.notify_all();`, and then hangs in the `t.join()` loop.

Now for whatever reason the worker thread get's its share of the CPU and continues running, and will run right into the `cond.wait(lock);` without checking `m_request_stop`. There it won't receive the notification because it came before it called the `wait()`, so the worker thread will now wait.

So now the master hangs in the `join()` loop, and the worker waits for notification => deadlock.

hebasto replied:
You're right. Going to amend the patch.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/26762#issuecomment-1500980698).

martinus commented on src/checkqueue.h:
```diff
@@ -155,6 +141,13 @@ class CCheckQueue
+    // pool `m_worker_threads`, copy operations are not appropriate.
+    CCheckQueue(const CCheckQueue&) = delete;
+    CCheckQueue& operator=(const CCheckQueue&) = delete;
+    CCheckQueue(CCheckQueue&&) = default;
```
I'd explicitly set the move operations to `= delete`. I think `std::mutex` is  not movable so these will be deleted anyways, but it feels safer to just mark them as deleted too. There will be threads running throughout the lifetime of `CCheckQueue`, so moving the object while something might be going on or threads waiting on a lock wouldn't be safe

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/26762#issuecomment-1500980698).

hebasto commented:
Updated f1370b2c1586f7fe487d9f17ee53bcd9b87a9f23 -> d462e3da7fc6be75269e88928fd80fc98c405474 ([pr26762.05](https://github.com/hebasto/bitcoin/commits/pr26762.05) -> [pr26762.07](https://github.com/hebasto/bitcoin/commits/pr26762.07), [diff](https://github.com/hebasto/bitcoin/compare/pr26762.05..pr26762.07)):

- addressed @martinus's comments

hebasto commented on src/checkqueue.h:
```diff
@@ -131,22 +129,10 @@ class CCheckQueue
-        }
-        assert(m_worker_threads.empty());
-        for (int n = 0; n < threads_num; ++n) {
+        for (int n = 0; n < worker_threads_num; ++n) {
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/26762#issuecomment-1500980698).

martinus reviewed (approved):
Code review ACK d462e3da7fc6be75269e88928fd80fc98c405474, as far as I can say now deadlocks are not possible any more when shutting down. I started an initial sync with `-assumevalid=0` so script checks are done, stopping the process a few times with Ctrl+C, all good; shutdown is still quick. `-par` still works too. Also did a rebase with `--exec "make -j18 check"`, so each commit individually builds & all tests work.

hebasto commented:
Rebased d462e3da7fc6be75269e88928fd80fc98c405474 -> 52129e335cbc68ac5d863f283f7d1a328ce79581 ([pr26762.07](https://github.com/hebasto/bitcoin/commits/pr26762.07) -> [pr26762.08](https://github.com/hebasto/bitcoin/commits/pr26762.08)) due to the conflict with #27419.

martinus commented:
re-ack 52129e335cbc68ac5d863f283f7d1a328ce79581

hebasto commented:
Rebased 52129e335cbc68ac5d863f283f7d1a328ce79581 -> 2076d846cc917cbafe61937a99b7867067011341 ([pr26762.08](https://github.com/hebasto/bitcoin/commits/pr26762.08) -> [pr26762.09](https://github.com/hebasto/bitcoin/commits/pr26762.09)) due to the conflict with #25977.

hebasto commented:
Rebased 2076d846cc917cbafe61937a99b7867067011341 -> b023c26eb8eda4c7f80ad2e7ebe1fb046e87d2ee ([pr26762.09](https://github.com/hebasto/bitcoin/commits/pr26762.09) -> [pr26762.10](https://github.com/hebasto/bitcoin/commits/pr26762.10)) due to the conflict with #27636.

hebasto committed:
Move global `scriptcheckqueue` into `ChainstateManager` class

hebasto committed:
Make `CCheckQueue` destructor stop worker threads

hebasto committed:
refactor: Make `CCheckQueue` constructor start worker threads

hebasto committed:
refactor: Drop unneeded declaration

hebasto committed:
refactor: Make `CCheckQueue` non-copyable and non-movable explicitly

hebasto committed:
refactor: Move `{MAX,DEFAULT}_SCRIPTCHECK_THREADS` constants

hebasto commented:
Rebased e89273e80765219ff545ee5c26a257aef7d69f9c -> b7ebc999ee373f0429939ac0bdb8e38fb37404aa ([pr26762.11](https://github.com/hebasto/bitcoin/commits/pr26762.11) -> [pr26762.12](https://github.com/hebasto/bitcoin/commits/pr26762.12)) due to the conflict with #27861.

hebasto commented:
Rebased b7ebc999ee373f0429939ac0bdb8e38fb37404aa -> 6e079015707188ac15405662ce396dd837da569a ([pr26762.12](https://github.com/hebasto/bitcoin/commits/pr26762.12) -> [pr26762.13](https://github.com/hebasto/bitcoin/commits/pr26762.13)) due to the conflict with #27607.

hebasto commented:
Rebased 6e079015707188ac15405662ce396dd837da569a -> 2235765f7fd946d3b08fd9e109584967407e46d1 ([pr26762.13](https://github.com/hebasto/bitcoin/commits/pr26762.13) -> [pr26762.14](https://github.com/hebasto/bitcoin/commits/pr26762.14)) due to the conflict with #28048.

martinus commented:
code review ACK 2235765f7fd946d3b08fd9e109584967407e46d1, also rebased & built & tested each individual commit with `make check`

hebasto commented:
Rebased 2235765f7fd946d3b08fd9e109584967407e46d1 -> 5b3ea5fa2e7f6dc1c9161ed8b74c9be4bd1e92dd ([pr26762.14](https://github.com/hebasto/bitcoin/commits/pr26762.14) -> [pr26762.15](https://github.com/hebasto/bitcoin/commits/pr26762.15)) due to the conflict with #27596.

cacrowley reviewed (commented):
#26762

sedited reviewed (approved):
ACK 5b3ea5fa2e7f6dc1c9161ed8b74c9be4bd1e92dd

martinus commented:
ACK 5b3ea5fa2e7

achow101 commented:
ACK 5b3ea5fa2e7f6dc1c9161ed8b74c9be4bd1e92dd

maflcko reviewed (commented):
lgtm
